### PR TITLE
MANIFEST.in and setup.py clean-up

### LINKED ISSUE #10335: package "tests" in Python path?
## 🐛 Bug

PyTorch Lightning install its tests into "test" in the Python path. I would venture that this is not ideal.

### To Reproduce

Install PyTorch Lightning
```
import pathlib
import pytorch_lightning
print(list((pathlib.Path(pytorch_lightning.__file__).parent.parent / 'tests').glob('*')))
```

This is with git a checkout from today.

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

- Do not include tests, pl_examples, and benchmarks in the generated wheel (`python setup.py bdist_wheel`)
- Prune the MANIFEST.in definition for the source distribution (`python setup.py sdist`)
- Copy the MNIST implementation from our tests into pl_examples
- Removed `check-manifest` use as it contradicts the output of `bdist_wheel`:

```console
warning: no previously-included files found matching '.yapfignore'
warning: no previously-included files found matching '*.yaml'
warning: no previously-included files found matching 'Makefile'
warning: no previously-included files matching '*.py' found under directory 'requirements'
no previously-included directories found matching 'docs'
no previously-included directories found matching 'dockers'
no previously-included directories found matching 'legacy'
no previously-included directories found matching 'benchmarks'
no previously-included directories found matching 'notebooks'
no previously-included directories found matching 'pl_examples'
no previously-included directories found matching 'tests'
```

Credit to @flying-sheep for finding the solution in #7596
Closes #7596
Fixes #10335 

Resources:
- https://setuptools.readthedocs.io/en/latest/userguide/package_discovery.html#using-find-or-find-packages
- https://packaging.python.org/guides/using-manifest-in

Manually tested that everything is correctly included/excluded.

Part of https://github.com/PyTorchLightning/internal-dev/issues/132

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [n/a] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [n/a] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

cc @borda @justusschock @awaelchli @akihironitta

### CODE REVIEW

File: .azure-pipelines/gpu-tests.yml [hunk 1]
```diff
@@ -123,6 +123,4 @@ jobs:
         # cd pl_examples/basic_examples
         # bash submit_ddp_job.sh
         # bash submit_ddp2_job.sh
-      env:
-        PL_USE_MOCKED_MNIST: "1"
```
Comment (Borda) [inline .azure-pipelines/gpu-tests.yml:127]: why do you remove it?

File: .github/workflows/ci_pkg-install.yml [hunk 2]
```diff
@@ -27,12 +27,11 @@ jobs:
 
       - name: Prepare env
         run: |
-          pip install check-manifest "twine==3.2" setuptools wheel
+          pip install "twine==3.2" setuptools wheel
 
       - name: Create package
         run: |
-          check-manifest
```
Comment (Borda) [inline .github/workflows/ci_pkg-install.yml:34]: why not check the manifest?

File: pl_examples/basic_examples/autoencoder.py [hunk 3]
```diff
@@ -32,8 +32,6 @@
     from torchvision import transforms
 if _TORCHVISION_MNIST_AVAILABLE:
     from torchvision.datasets import MNIST
-else:
-    from tests.helpers.datasets import MNIST
```
Comment (Borda) [inline pl_examples/basic_examples/autoencoder.py:36]: so it will fail if MNIST cannot be downloaded?
cc: @awaelchli

File: .github/workflows/ci_pkg-install.yml (hunk 2 above)
Comment (carmocca) [reply .github/workflows/ci_pkg-install.yml:34]: Because the warnings it outputs doesn't seem to be correct, if you see the changes to `MANIFEST.in`, we just need to include two things (everything else gets included/excluded automatically)

I checked by looking at the generated distribution.

However, `check-manifest` says all the includes/excludes are necessary but doesn't seem like it.
I don't understand why exactly

Comment (carmocca) [reply .github/workflows/ci_pkg-install.yml:34]: Also, have a look at the PR description

File: pl_examples/basic_examples/autoencoder.py (hunk 3 above)
Comment (carmocca) [reply pl_examples/basic_examples/autoencoder.py:36]: So I tried removing this because we are in a deadlock

On one hand, we want to exclude the tests directory from the distributed packages
But on the other hand, we need the test MNIST helpers to run some of the examples

So we have 4 options:

1. Do not exclude tests from the distribution
2. Do not use our own MNIST implementation (this is what I explored with the current changes)
3. Move the MNIST implementation to the `pytorch_lightning` folder
4. Copy the MNIST implementation in the `example` file that uses it.

@Borda how do you think we should proceed?

File: .azure-pipelines/gpu-tests.yml (hunk 1 above)
Comment (carmocca) [reply .azure-pipelines/gpu-tests.yml:127]: Answering below, this is WIP

File: pl_examples/basic_examples/autoencoder.py (hunk 3 above)
Comment (tchaton) [reply pl_examples/basic_examples/autoencoder.py]: I think the 3 or 4 are fine, but I would prefer 4.

Comment (Borda) [reply pl_examples/basic_examples/autoencoder.py]: I think the best is 4)

Comment (justusschock) [reply pl_examples/basic_examples/autoencoder.py]: I would be against 3). This is nothing related to PL itself.
I also think 4) is best.

**Review by tchaton — APPROVED** [review]
LGTM !

File: tests/helpers/datasets.py [hunk 4]
```diff
@@ -19,7 +19,6 @@
 from typing import Optional, Sequence, Tuple
 
 import torch
-from torch import Tensor
```
Comment (awaelchli) [inline tests/helpers/datasets.py:22]: why is this wrong?

File: .azure-pipelines/gpu-benchmark.yml [hunk 5]
```diff
@@ -36,7 +36,7 @@ jobs:
 
     steps:
       - bash: |
-          python -m pytest benchmarks -v --durations=0
+          python -m pytest tests/benchmarks -v --durations=0
```
Comment (awaelchli) [inline .azure-pipelines/gpu-benchmark.yml:39]: Is it not possible to keep the benchmarks folder separate?

Comment (carmocca) [reply .azure-pipelines/gpu-benchmark.yml:39]: Nope

File: tests/helpers/datasets.py (hunk 4 above)
Comment (carmocca) [reply tests/helpers/datasets.py:22]: Just making sure both implementations match exactly

File: .azure-pipelines/gpu-benchmark.yml (hunk 5 above)
Comment (Borda) [reply .azure-pipelines/gpu-benchmark.yml:39]: in such case we shall rather make it stand-alone and eventually outsource in compatibility repo

Comment (carmocca) [reply .azure-pipelines/gpu-benchmark.yml:39]: Sure, it's probably going to be changed after the community repo is fully formalized and our benchmarks review.

It can be done as a follow-up

**Review by awaelchli — APPROVED** [review]

File: MANIFEST.in [hunk 6]
```diff
@@ -11,69 +11,8 @@
...
-# Manifest syntax https://docs.python.org/2/distutils/sourcedist.html
-graft wheelhouse
-
-recursive-exclude __pycache__  *.py[cod] *.orig
-
-# Include the README and CHANGELOG
-include *.md
-
-# Include the license file
-include LICENSE
```
Comment (akihironitta) [inline MANIFEST.in:24]: Is it okay to drop the license file as well?

Comment (carmocca) [reply MANIFEST.in:24]: It is included automatically, you can check with

`python setup.py sdist`

**Review by akihironitta — APPROVED** [review]

**Review by Borda — DISMISSED** [review]
pls make the benchmark move to test in separate PR, as it does not have any connection to Manifest/setup cleaning

**Review by Borda — APPROVED** [review]
